Update haproxy best practices using haproxy2.5.0 (#7046)
pingcap/docs: best-practices/haproxy-best-practices.md

HAProxy is free, open-source software written in C language that provides a high availability load balancer and proxy server for TCP and HTTP-based applications. Because of its fast and efficient use of CPU and memory, HAProxy is now widely used by many well-known websites such as GitHub, Bitbucket, Stack Overflow, Reddit, Tumblr, Twitter, Tuenti, and AWS (Amazon Web Services).

> HAProxy is written in the year 2000 by Willy Tarreau, the core contributor to the Linux kernel, who is still responsible for the maintenance of the project and provides free software updates in the open-source community. The latest stable version 2.5.0 was released on November 23, 2021, bringing more [excellent features](https://www.haproxy.com/blog/announcing-haproxy-2-5/).

Reviewer: ```suggestion
HAProxy is written in the year 2000 by Willy Tarreau, the core contributor to the Linux kernel, who is still responsible for the maintenance of the project and provides free software updates in the open-source community. In this guide, HAProxy [2.5.0](https://www.haproxy.com/blog/announcing-haproxy-2-5/) is used. It is recommended to use the latest stable version. See [the released version of HAProxy](http://www.haproxy.org/) for details.
```

## Deploy HAProxy

You can easily use HAProxy to configure and set up a load-balanced database environment. This section shows general deployment operations. You can customize the [configuration file](http://cbonte.github.io/haproxy-dconv/2.5/configuration.html) based on your actual scenario.

### Install HAProxy

1. Download the package of the HAProxy 2.5.0 source code:

    {{< copyable "shell-regular" >}}

    ```bash
    wget https://github.com/haproxy/haproxy/archive/refs/tags/v2.5.0.zip
    ```

2. Unzip the package:

    {{< copyable "shell-regular" >}}

    ```bash
    unzip v2.5.0.zip
    ```

3. Compile the application from the source code:

    {{< copyable "shell-regular" >}}

    ```bash
    cd haproxy-2.5.0
    make clean

>     make -j 8 TARGET=linux-glibc  USE_THREAD=1

Reviewer 2: please remove the redundant space

```
make -j 8 TARGET=linux-glibc USE_THREAD=1
```

Reviewer 3: Thanks! The redundant space is removed in https://github.com/pingcap/docs/pull/7082.

>     make PREFIX=${/app/haproxy} SBINDIR=${/app/haproxy/bin} install

Reviewer: ```suggestion
    make PREFIX=${/app/haproxy} SBINDIR=${/app/haproxy/bin} install  # Replace `${/app/haproxy}` and `${/app/haproxy/bin}` with your custom directories.
```

## Deploy HAProxy

You can easily use HAProxy to configure and set up a load-balanced database environment. This section shows general deployment operations. You can customize the [configuration file](http://cbonte.github.io/haproxy-dconv/2.5/configuration.html) based on your actual scenario.

### Install HAProxy

> 1. Download HAProxy 2.5.0：
>    wget https://github.com/haproxy/haproxy/archive/refs/tags/v2.5.0.zip
> 2. unzip haproxy

Reviewer 3: ```suggestion
1. Download the package of the HAProxy 2.5.0 source code:

    {{< copyable "shell-regular" >}}

    ```bash
    wget https://github.com/haproxy/haproxy/archive/refs/tags/v2.5.0.zip
    ```

2. Unzip the package to get the source code:
```

    {{< copyable "shell-regular" >}}

    ```bash
    unzip v2.5.0.zip
    ```

> 3. make & make install

Reviewer 3: ```suggestion
3. Compile the application from the source code:
```

>      {{< copyable "shell-regular" >}}
>     ```bash
>     cd haproxy-2.5.0
>     make clean
>     make -j 8 TARGET=linux-glibc  USE_THREAD=1
>     make PREFIX=${/app/haproxy} SBINDIR=${/app/haproxy/bin} install
>     ```

Reviewer 3: ```suggestion

    {{< copyable "shell-regular" >}}

    ```bash
    cd haproxy-2.5.0
    make clean
    make -j 8 TARGET=linux-glibc  USE_THREAD=1
    make PREFIX=${/app/haproxy} SBINDIR=${/app/haproxy/bin} install
    ```

```

> 4. reconfig profile

Reviewer 3: ```suggestion
4. Reconfigure the profile:
```

>      {{< copyable "shell-regular" >}}
>     ```bash
>     echo 'export PATH=/app/haproxy/bin:$PATH' >> /etc/profile
>     ```

Reviewer 3: ```suggestion
    {{< copyable "shell-regular" >}}

    ```bash
    echo 'export PATH=/app/haproxy/bin:$PATH' >> /etc/profile
    ```
```

## Deploy HAProxy

You can easily use HAProxy to configure and set up a load-balanced database environment. This section shows general deployment operations. You can customize the [configuration file](http://cbonte.github.io/haproxy-dconv/2.5/configuration.html) based on your actual scenario.

### Install HAProxy

1. Download the package of the HAProxy 2.5.0 source code:

    {{< copyable "shell-regular" >}}

    ```bash
    wget https://github.com/haproxy/haproxy/archive/refs/tags/v2.5.0.zip
    ```

2. Unzip the package:

    {{< copyable "shell-regular" >}}

    ```bash
    unzip v2.5.0.zip
    ```

3. Compile the application from the source code:

    {{< copyable "shell-regular" >}}

    ```bash
    cd haproxy-2.5.0
    make clean
    make -j 8 TARGET=linux-glibc  USE_THREAD=1
    make PREFIX=${/app/haproxy} SBINDIR=${/app/haproxy/bin} install
    ```

> 4. Reconfigure the profile:
>     {{< copyable "shell-regular" >}}

Reviewer: ```suggestion
4. Reconfigure the profile:

    {{< copyable "shell-regular" >}}
```

    ```bash
    echo 'export PATH=/app/haproxy/bin:$PATH' >> /etc/profile
    ```

> 5. Check whether the installation is successful：

Reviewer: ```suggestion
5. Check whether the installation is successful:
```

    {{< copyable "shell-regular" >}}

    ```bash
    which haproxy
    ```

> 
> 

Reviewer: Delete an extra blank line.
```suggestion

```

   log         127.0.0.1 local2            # Global syslog servers (up to two).
   chroot      /var/lib/haproxy            # Changes the current directory and sets superuser privileges for the startup process to improve security.
   pidfile     /var/run/haproxy.pid        # Writes the PIDs of HAProxy processes into this file.
   maxconn     256                         # The maximum number of concurrent connections per HAProxy thread.

>    nbthread    48                          # The maximum number of threads. （The upper limit is equal to the number of cpus)

Reviewer: ```suggestion
   nbthread    48                          # The maximum number of threads. (The upper limit is equal to the number of CPUs)
```

### Start HAProxy

> Execute `haproxy`. `/etc/haproxy/haproxy.cfg` is read by default (recommended).

Reviewer 3: ```suggestion
To start HAProxy, run `haproxy`. `/etc/haproxy/haproxy.cfg` is read by default (recommended).
```

{{< copyable "shell-regular" >}}

```bash
haproxy -f /etc/haproxy/haproxy.cfg
```

### Stop HAProxy

> Use `kill -9`.

Reviewer 3: ```suggestion
To stop HAProxy, use the `kill -9` command.
```

### Start HAProxy

To start HAProxy, run `haproxy`. `/etc/haproxy/haproxy.cfg` is read by default (recommended).

{{< copyable "shell-regular" >}}

```bash
haproxy -f /etc/haproxy/haproxy.cfg
```

> 

Reviewer 3: ```suggestion
```
